examples/step-3/doc/intro.dox

```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
+The problem with this method is that it requires a number of operations that
+is proportional to $N^3$, where $N$ is the number of equations or unknowns
+in the linear system -- more specifically, the number of operations is
+$\frac 23 N^3$, give or take a few. With $N=1089$, this means that we would
+have to do around $861$ million operations. This is a number that is quite
+feasible and it would take modern processors on the order of 1-10 seconds
+to do this. But it is clear that this isn't going to scale: If we have ten times
+as many equations in the linear system (that is, ten times as many unknowns),
+then it would already take 1000-10,000 seconds or on the order of an hour. Make
+the linear system another ten times larger, and it is clear that we can not solve
+it any more on a single computer.
+
+One can rescue the situation somewhat by realizing that only a relatively
+small number of entries in the matrix are nonzero -- that is, the matrix
+is [sparse](https://en.wikipedia.org/wiki/Sparse_matrix). Variations of
+Gaussian elimination can exploit this, making the process substantially
+faster; we will use one such method -- implemented in the SparseDirectUMFPACK
+class -- in step-29 for the first time, among several others than come after
+that. These variations of Gaussian elimination might get us to problem
+sizes on the order of 100,000 or 200,000, but not all that much beyond that.
+
+Instead, what we will do here is take up an idea from 1952: the
+[Conjugate Gradient method](https://en.wikipedia.org/wiki/Conjugate_gradient_method),
+or in short "CG". CG is an "iterative" solver in that it forms a sequence
+of vectors that *converge* to the exact solution; in fact, after $N$ such iterations
+it finds the exact solution if the matrix is symmetric and positive definite.
+The method was originally developed as another way
+to solve a linear system exactly, like Gaussian elimination, but as such
+it had few advantages and was largely forgotten for a few decades. But,
+when computers became powerful enough to solve problems of a size where
+Gaussian elimination doesn't work well any more (sometime in the 1980s),
+CG was rediscovered as a way to people realized that it is well suited for
+large and sparse systems like the ones we get from the finite element method.
+This is because (i) the vectors it computes *converge* to the exact solution,
+and consequently we do not actually have to do all $N$ iterations to
+find the exact solution as long as we're happy with reasonably good approximations;
+and (ii) it only ever requires matrix-vector products, which is very useful
+for sparse matrices because a sparse matrix has, by definition, only
+${\cal O}(N)$ entries and so a matrix-vector product can be done with
+${\cal O}(N)$ effort whereas it costs $N^2$ operations to do the same for
+dense matrices. As a consequence, we can hope to solve linear systems
+with at most ${\cal O}(N^2)$ operations, and in some cases substantially
```

Comment:
```suggestion
with at most ${\cal O}(N^2)$ operations, and in many cases substantially
```
Poisson-type problems (which is one of the few cases where plain CG actually works) have a cost of O(N^{3/2}) in 2D and O(N^{4/3}) in 3D. The reason for the different powers is that in 3D, for the same N ~ h^{-3}, the ratio between the domain size and cell length is lower.

Reply:
I assume you were talking about direct solvers:
Just for completeness, I think it's O(N^{5/3}) in 3d -- worse than the 3/2 in 2d. In 2d, you can re-sort DoFs in such a way that the bandwidth of the matrix is b=O(sqrt(N)), but in 3d the bandwidth becomes b=O(N^{2/3}).